modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyAllocator.java

```diff
@@ -79,12 +110,30 @@ public static ByteBufAllocator getAllocator() {
         }
     }
 
-    private static boolean useUnpooled() {
+    private static boolean useUnpooled(long heapSizeInBytes, long g1RegionSize) {
+        if (userForcedUnpooled()) {
+            return true;
+        } else {
+            boolean heapIsOneGBOrLess = heapSizeInBytes <= 1 << 30;
+            boolean g1gcRegionIsLessThan1MB = g1RegionSize < 1 << 20;
```

Comment:
region size cannot be smaller than 1MB, so hopefully this never triggers. I suppose a future JDK could lower the minimum region size so OK to keep the check, but I think we should add a comment about this. It is hard to assert anything meaningful about in a test without spinning up a small separate JVM.

You could also change the `maxOrder` calculation to be something like `log2(g1RegionSize / pageSize)` and trigger unpooled when smaller than 5. Not necessary though, current code is fine too with a comment.